codex/node.nim

```diff
@@ -43,14 +43,18 @@ type
 
   CodexError = object of CatchableError
 
+  Contracts* = tuple
+    client: ?ClientInteractions
+    host: ?HostInteractions
```

Comment:
Another thing to keep in mind is that Nim always initialized object types, so there is no way of telling wether they were set or not, in this cases optionals are used to signal that. With ref types this isn't an issue because they are nillable, but pure object types aren't.

In other words, if you have code like this:

```nim
type
  Obj1 = object
     val: int

  Obj2 = object
     val: string
  
  Obj3 = object
    o1: Obj1
    o2: Obj2

let obj3 = Obj3(); # you can't tell wether o1 and o2 where initialized or not

# try doing something if o1 or o2 are set - you can't, o1 > 0 and o2.len > 0, don't really signal wether they were set

  Obj4 = object
    o1: ?Obj1
    o2: ?Obj2


let obj4 = Obj4(o1: 0.some, 02: "".some) 

  # you can now do 
  if obj4.o1.isSome: 
    echo "Yay, o1 was set!"
  else:
    echo "Yaiks, o1 was not set"
```

Reply:
> Out of curiosity, why do you define `client` & `host` as optional? Is there a possible case where one will be defined and the other not? From top of my head, I can't think of any.

A codex node can play different roles in the network, `host` and `client` being two of them. In the `host` role, the node will offer storage for sale, and only a small portion of nodes in the network will do that. In the `client` role nodes will buy storage, which we expect a larger portion of the nodes in the network to do. But the vast majority of nodes will probably only download content from the network, which requires neither the `host` nor the `client` role.

Reply:
@markspanbroek makes sense, but currently, there is no way how to specify if the node wants to be host or client right? (That is why I have pointed this out) So this is envisioned to be configurable in the future?

Reply:
Good point @AuHau, you currently cannot specify this on the command line. I do want to keep the option open in the architecture though. We're currently starting things like `Proving` and `Sales`, even when we don't have any storage available to sell. Which is fine for now, it's harmless. But I'd like to have the option in the future to only start this machinery when needed.